File: src/python_examples/cgalpy_examples/Polyhedron_3/pol3_visual_buffer_raster.cpp
// Author(s): Utkarsh Khajuria <utkarshkhajuria55@gmail.com>

#include <CGAL/Exact_predicates_exact_constructions_kernel.h>
#include <CGAL/Graphics_scene.h>
#include <CGAL/Polyhedron_3.h>
#include <CGAL/draw_polyhedron.h>

#include <cmath>
#include <cstddef>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace {

using Kernel = CGAL::Exact_predicates_exact_constructions_kernel;
using Point = Kernel::Point_3;
using Polyhedron = CGAL::Polyhedron_3<Kernel>;

constexpr int width = 192;
constexpr int height = 192;
constexpr int margin = 12;

constexpr double u_min = -0.50;
constexpr double u_max = 1.05;
constexpr double v_min = -0.05;
constexpr double v_max = 1.05;

std::pair<int, int> project(double x, double y, double z) {
  const double u = x - 0.45 * y;
  const double v = z + 0.35 * x + 0.35 * y;

  if (u < u_min || u > u_max || v < v_min || v > v_max) {
    throw std::runtime_error(
      "projected point outside fixed viewport");
  }

  const int px = margin + static_cast<int>(std::floor(
    ((u - u_min) / (u_max - u_min))
      * static_cast<double>(width - 1 - 2 * margin) + 0.5));

  const int py = height - 1 - margin - static_cast<int>(std::floor(
    ((v - v_min) / (v_max - v_min))
      * static_cast<double>(height - 1 - 2 * margin) + 0.5));

  return {px, py};
}

void set_black(std::vector<unsigned char>& image, int x, int y) {
  if (x < 0 || x >= width || y < 0 || y >= height) {
    throw std::runtime_error("pixel outside raster");
  }

  const std::size_t offset =
    static_cast<std::size_t>((y * width + x) * 3);
  image[offset] = 0;
  image[offset + 1] = 0;
  image[offset + 2] = 0;
}

void draw_line(std::vector<unsigned char>& image,
               int x0, int y0, int x1, int y1) {
  const int dx = std::abs(x1 - x0);
  const int sx = x0 < x1 ? 1 : -1;
  const int dy = -std::abs(y1 - y0);
  const int sy = y0 < y1 ? 1 : -1;
  int error = dx + dy;

  while (true) {
    set_black(image, x0, y0);
    if (x0 == x1 && y0 == y1) {
      break;
    }

    const int twice_error = 2 * error;
    if (twice_error >= dy) {
      error += dy;
      x0 += sx;
    }
    if (twice_error <= dx) {
      error += dx;
      y0 += sy;
    }
  }
}

void write_ppm(const std::string& path,
               const std::vector<unsigned char>& image) {
  std::ofstream output(path, std::ios::binary);
  if (!output) {
    throw std::runtime_error("unable to open raster output");
  }

  output << "P6\n" << width << ' ' << height << "\n255\n";
  output.write(
    reinterpret_cast<const char*>(image.data()),
    static_cast<std::streamsize>(image.size()));

  if (!output) {
    throw std::runtime_error("failed while writing raster output");
  }
}

} // namespace

int main(int argc, char** argv) {
  try {
    if (argc != 2) {
      std::cerr
        << "usage: pol3_visual_buffer_raster <output.ppm>\n";
      return 2;
    }

    Polyhedron polyhedron;
    polyhedron.make_tetrahedron(
      Point(1, 0, 0),
      Point(0, 1, 0),
      Point(0, 0, 1),
      Point(0, 0, 0));

    CGAL::Graphics_scene scene;
    CGAL::add_to_graphics_scene(polyhedron, scene);

    const auto& buffer =
      scene.get_array_of_index(CGAL::Graphics_scene::POS_SEGMENTS);

    if (buffer.empty()) {
      throw std::runtime_error(
        "Polyhedron graphics scene has no POS_SEGMENTS");
    }
    if (buffer.size() % 6 != 0) {
      throw std::runtime_error(
        "POS_SEGMENTS float count is not divisible by 6");
    }

    std::vector<unsigned char> image(
      static_cast<std::size_t>(width * height * 3),
      static_cast<unsigned char>(255));

    for (std::size_t offset = 0; offset < buffer.size(); offset += 6) {
      const auto p0 = project(
        buffer[offset],
        buffer[offset + 1],
        buffer[offset + 2]);
      const auto p1 = project(
        buffer[offset + 3],
        buffer[offset + 4],
        buffer[offset + 5]);

      draw_line(
        image,
        p0.first,
        p0.second,
        p1.first,
        p1.second);
    }

    write_ppm(argv[1], image);
    std::cout << "raster_width " << width << '\n';
    std::cout << "raster_height " << height << '\n';
    std::cout << "raster_segments " << buffer.size() / 6 << '\n';
    return 0;
  }
  catch (const std::exception& error) {
    std::cerr << error.what() << '\n';
    return 1;
  }
}

File: src/python_examples/cgalpy_examples/Polyhedron_3/pol3_visual_buffer_raster.py
#!/usr/bin/python
# Author(s): Utkarsh Khajuria <utkarshkhajuria55@gmail.com>

import importlib
import math
import sys
from pathlib import Path

WIDTH = 192
HEIGHT = 192
MARGIN = 12

U_MIN = -0.50
U_MAX = 1.05
V_MIN = -0.05
V_MAX = 1.05


def project(x, y, z):
    u = x - 0.45 * y
    v = z + 0.35 * x + 0.35 * y
    if not (U_MIN <= u <= U_MAX and V_MIN <= v <= V_MAX):
        raise ValueError(
            f"projected point outside fixed viewport: ({u}, {v})"
        )

    px = MARGIN + int(
        math.floor(
            ((u - U_MIN) / (U_MAX - U_MIN))
            * (WIDTH - 1 - 2 * MARGIN)
            + 0.5
        )
    )
    py = HEIGHT - 1 - MARGIN - int(
        math.floor(
            ((v - V_MIN) / (V_MAX - V_MIN))
            * (HEIGHT - 1 - 2 * MARGIN)
            + 0.5
        )
    )
    return px, py


def set_black(image, x, y):
    if not (0 <= x < WIDTH and 0 <= y < HEIGHT):
        raise ValueError(f"pixel outside raster: ({x}, {y})")

    offset = (y * WIDTH + x) * 3
    image[offset] = 0
    image[offset + 1] = 0
    image[offset + 2] = 0


def draw_line(image, x0, y0, x1, y1):
    dx = abs(x1 - x0)
    sx = 1 if x0 < x1 else -1
    dy = -abs(y1 - y0)
    sy = 1 if y0 < y1 else -1
    error = dx + dy

    while True:
        set_black(image, x0, y0)
        if x0 == x1 and y0 == y1:
            break

        twice_error = 2 * error
        if twice_error >= dy:
            error += dy
            x0 += sx
        if twice_error <= dx:
            error += dx
            y0 += sy


def write_ppm(path, image):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(
        f"P6\n{WIDTH} {HEIGHT}\n255\n".encode("ascii")
        + bytes(image)
    )


if len(sys.argv) != 3:
    raise SystemExit(
        "usage: pol3_visual_buffer_raster.py "
        "<binding-library> <output.ppm>"
    )

library = sys.argv[1]
output = Path(sys.argv[2])

CGALPY = importlib.import_module(library)
Ker = CGALPY.Ker
Pol3 = CGALPY.Pol3
Bvr = CGALPY.Bvr

polyhedron = Pol3.Polyhedron_3()
polyhedron.make_tetrahedron(
    Ker.Point_3(1, 0, 0),
    Ker.Point_3(0, 1, 0),
    Ker.Point_3(0, 0, 1),
    Ker.Point_3(0, 0, 0),
)

scene = Bvr.Graphics_scene()
Bvr.add_to_graphics_scene(polyhedron, scene)

buffer = scene.buffer(Bvr.POS_SEGMENTS)
if not buffer:
    raise RuntimeError("Polyhedron graphics scene has no POS_SEGMENTS")
if len(buffer) % 6 != 0:
    raise RuntimeError(
        f"POS_SEGMENTS float count is not divisible by 6: {len(buffer)}"
    )

image = bytearray([255]) * (WIDTH * HEIGHT * 3)

for offset in range(0, len(buffer), 6):
    x0, y0, z0, x1, y1, z1 = buffer[offset:offset + 6]
    p0 = project(float(x0), float(y0), float(z0))
    p1 = project(float(x1), float(y1), float(z1))
    draw_line(image, p0[0], p0[1], p1[0], p1[1])

write_ppm(output, image)
print("raster_width", WIDTH)
print("raster_height", HEIGHT)
print("raster_segments", len(buffer) // 6)
